qemu/qemu.go

```diff
@@ -26,14 +26,616 @@ package qemu
+// It allows for passing custom settings and parameters to the qemu API.
+type Config struct {
+	// Path is the qemu binary path.
+	Path string
+
+	// Ctx is not used at the moment.
+	Ctx context.Context
+
+	// Name is the qemu guest name
+	Name string
+
+	// UUID is the qemu process UUID.
+	UUID string
+
+	// CPUModel is the CPU model to be used by qemu.
+	CPUModel string
+
+	// Machine
+	Machine Machine
+
+	// QMPSocket is the QMP socket description.
+	QMPSocket QMPSocket
+
+	// Devices is a list of devices for qemu to create.
+	Devices []Device
+
+	// NetDevices is a list of networking devices for qemu to create.
+	NetDevices []NetDevice
+
+	// CharDevices is a list of character devices for qemu to export.
+	CharDevices []string
+
+	// Objects is a list of objects for qemu to create.
+	Objects []Object
+
+	// FilesystemDevices is a list of filesystem devices.
+	FilesystemDevices []FSDevice
+
+	// RTC is the qemu Real Time Clock configuration
+	RTC RTC
+
+	// VGA is the qemu VGA mode.
+	VGA string
+
+	// Kernel is the guest kernel configuration.
+	Kernel Kernel
+
+	// Memory is the guest memory configuration.
+	Memory Memory
+
+	// SMP is the quest multi processors configuration.
+	SMP SMP
+
+	// GlobalParam is the -global parameter.
+	GlobalParam string
+
+	// Knobs is a set of qemu boolean settings.
+	Knobs Knobs
+
+	// FDs is a list of open file descriptors to be passed to the spawned qemu process
+	FDs []*os.File
+}
+
+func appendName(params []string, config Config) []string {
+	if config.Name != "" {
+		params = append(params, "-name")
+		params = append(params, config.Name)
+	}
+
+	return params
+}
+
+func appendMachine(params []string, config Config) []string {
+	if config.Machine.Type != "" {
+		var machineParams []string
+
+		machineParams = append(machineParams, config.Machine.Type)
+
+		if config.Machine.Acceleration != "" {
+			machineParams = append(machineParams, fmt.Sprintf(",accel=%s", config.Machine.Acceleration))
+		}
+
+		params = append(params, "-machine")
+		params = append(params, strings.Join(machineParams, ""))
+	}
+
+	return params
+}
+
+func appendCPUModel(params []string, config Config) []string {
+	if config.CPUModel != "" {
+		params = append(params, "-cpu")
+		params = append(params, config.CPUModel)
+	}
+
+	return params
+}
+
+func appendQMPSocket(params []string, config Config) []string {
+	if config.QMPSocket.Type != "" && config.QMPSocket.Name != "" {
+		var qmpParams []string
+
+		qmpParams = append(qmpParams, fmt.Sprintf("%s:", config.QMPSocket.Type))
+		qmpParams = append(qmpParams, fmt.Sprintf("%s", config.QMPSocket.Name))
+		if config.QMPSocket.Server == true {
+			qmpParams = append(qmpParams, ",server")
+			if config.QMPSocket.NoWait == true {
+				qmpParams = append(qmpParams, ",nowait")
+			}
+		}
+
+		params = append(params, "-qmp")
+		params = append(params, strings.Join(qmpParams, ""))
+	}
+
+	return params
+}
+
+func appendDevices(params []string, config Config) []string {
```

Comment:
@sameo appending devices in some cases may need appending to the kernel command line also. So we will probably need the ability to pass out any additional kernel params generated by the device append.

Note: The comment applies to other devices too like char devices and fs devices